### PR TITLE
improved implementation of AddBraiding*

### PR DESCRIPTION
**Pull request by mohamed-barakat**
by replacing

HomalgMatrixListList( PermutationMat( PermList( permlist, n ) ), R ) -> CertainRows( HomalgIdentityMatrix( n, R ), permlist )

improving the runtime for creating big external matrices by orders of magnitude

### CODE REVIEW

File: CompilerForCAP/gap/InferDataTypes.gi [hunk 1]
```diff
@@ -897,7 +897,7 @@ CapJitAddTypeSignature( "QUO_INT", [ IsInt, IsInt ], IsInt );
 CapJitAddTypeSignature( "IsZero", [ IsInt ], IsBool );
 CapJitAddTypeSignature( "OneImmutable", [ IsIntegers ], IsInt );
 CapJitAddTypeSignature( "^", [ IsPerm, IsInt ], IsPerm );
-CapJitAddTypeSignature( "PermList", [ IsList ], IsPerm );
+#CapJitAddTypeSignature( "PermList", [ IsList ], IsPerm );
```
Comment (zickgraf) [inline CompilerForCAP/gap/InferDataTypes.gi:900]: I suggest we keep this: We know that it works and `PermList` will not change anyway.

File: GradedModulePresentationsForCAP/gap/GradedModulePresentations.gi [hunk 2]
```diff
@@ -1274,10 +1272,8 @@ end );
...
 #         
-#         permutation_matrix := PermutationMat( 
-#                                 PermList( List( [ 1 .. rank ], i -> ( RemInt( i - 1, rank_2 ) * rank_1 + QuoInt( i - 1, rank_2 ) + 1 ) ) ),
-#                                 rank 
-#                               );
+#         permutation_matrix := CertainColumns( HomalgIdentityMatrix( rank, underlying_graded_ring ),
+#                                       List( [ 1 .. rank ], i -> ( RemInt( i - 1, rank_2 ) * rank_1 + QuoInt( i - 1, rank_2 ) + 1 ) ) );
 #         
 #         return PresentationMorphism( object_1_tensored_object_2,
 #                                      TransposedMatrix( HomalgMatrix( permutation_matrix, rank, rank, homalg_ring ) ), ## transposed of the left case
```
Comment (zickgraf) [inline GradedModulePresentationsForCAP/gap/GradedModulePresentations.gi:1279]:
```suggestion
#                                     permutation_matrix,
```

File: ModulePresentationsForCAP/gap/ModulePresentationsForCAP.gi [hunk 3]
```diff
@@ -908,13 +906,11 @@ InstallGlobalFunction( ADD_BRAIDING_RIGHT,
...
-        permutation_matrix := PermutationMat( 
-                                PermList( List( [ 1 .. rank ], i -> ( RemInt( i - 1, rank_2 ) * rank_1 + QuoInt( i - 1, rank_2 ) + 1 ) ) ),
-                                rank 
-                              );
+        permutation_matrix := CertainColumns( HomalgIdentityMatrix( rank, homalg_ring ),
+                                      List( [ 1 .. rank ], i -> ( RemInt( i - 1, rank_2 ) * rank_1 + QuoInt( i - 1, rank_2 ) + 1 ) ) );
         
         return PresentationMorphism( cat, object_1_tensored_object_2,
-                                     TransposedMatrix( HomalgMatrix( permutation_matrix, rank, rank, homalg_ring ) ), ## transposed of the left case
+                                     permutation_matrix, ## transposed of the left case
```
Comment (zickgraf) [inline ModulePresentationsForCAP/gap/ModulePresentationsForCAP.gi:913]:
```suggestion
                                     permutation_matrix,
```

File: GradedModulePresentationsForCAP/gap/GradedModulePresentations.gi (hunk 2 above)
Comment (mohamed-barakat) [reply GradedModulePresentationsForCAP/gap/GradedModulePresentations.gi]: Done.

File: CompilerForCAP/gap/InferDataTypes.gi (hunk 1 above)
Comment (mohamed-barakat) [reply CompilerForCAP/gap/InferDataTypes.gi]: Done.

File: ModulePresentationsForCAP/gap/ModulePresentationsForCAP.gi (hunk 3 above)
Comment (mohamed-barakat) [reply ModulePresentationsForCAP/gap/ModulePresentationsForCAP.gi]: Done.

**Pull request merged**